app/usecase/keys/consumer_cached.go

```diff
@@ -0,0 +1,81 @@
+package keys
+
+import (
+	"errors"
+)
+
+var _ Consumer = (*ConsumerCached)(nil)
+
+type bufferEntry struct {
+	key string
+	err error
+}
+
+type ConsumerCached struct {
+	delegate   Consumer
+	bufferSize int
+	buffer     chan bufferEntry
+}
+
+func (p ConsumerCached) ConsumeInBatch(maxCount uint) ([]string, error) {
+	keys := make([]string, 0, maxCount)
+
+	for ; maxCount > 0; maxCount-- {
+		// there is probability to get the list of keys with the size less than bufferSize
+		// in this case we listen to done channel to perform loop break
+		done := make(chan bool)
+
+		// we should load new keys as soon as our buffer is empty
+		if len(p.buffer) == 0 {
+			go func() {
+				p.loadKeys()
+				done <- true
+			}()
+		}
+
+		select {
```

Comment:
Should there be another loop outside `select`?

Reply:
Let me think again about the implementation. Maybe I will find a better solution... 
If you have some idea, please say to me

Reply:
Let me think again about the implementation. Maybe I will find a better solution... 
If you have some idea, please say to me

Reply:
If we do the loop, how many elements should we get from the channel?  It seems to me that the current approach handles it in the right way.

Reply:
@alldroll I read the code again. It seem that I misunderstood it before.